File: src/main/java/net/hectus/neobb/shop/Shop.java
package net.hectus.neobb.shop;

import com.marcpg.libpg.lang.Translation;
import net.hectus.neobb.NeoBB;
import net.hectus.neobb.lore.ItemLoreBuilder;
import net.hectus.neobb.player.NeoPlayer;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.util.Colors;
import org.bukkit.World;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public abstract class Shop {
    public final NeoPlayer player;
    public final ItemLoreBuilder loreBuilder;

    public final List<? extends Turn<?>> dummyTurns;

    public Shop(@NotNull NeoPlayer player) {
        this.player = player;
        this.dummyTurns = player.game.info().turns().stream().map(t -> turn(t, player)).toList();
        try {
            this.loreBuilder = player.game.info().loreBuilder().getConstructor().newInstance();
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    public static Turn<?> turn(@NotNull Class<? extends Turn<?>> clazz, NeoPlayer player) {
        try {
            if (clazz.getName().toLowerCase().contains("warp")) {
                return clazz.getConstructor(World.class).newInstance(player.game.world());
            } else {
                return clazz.getConstructor(NeoPlayer.class).newInstance(player);
            }
        } catch (ReflectiveOperationException e) {
            NeoBB.LOG.warn("Could not get turn!", e);
            return Turn.DUMMY;
        }
    }

    public abstract void open();

    public final void done() {
        player.inventory.setShopDone(true);
        if (player.game.allShopDone()) {
            player.game.start();
        } else {
            player.sendMessage(Translation.component(player.locale(), "shop.wait").color(Colors.EXTRA));
        }
    }
}

File: src/main/kotlin/net/hectus/neobb/shop/Shop.kt
package net.hectus.neobb.shop

import net.hectus.neobb.lore.ItemLoreBuilder
import net.hectus.neobb.player.NeoPlayer
import net.hectus.neobb.structure.PlacedStructure
import net.hectus.neobb.turn.Turn
import net.hectus.neobb.util.Colors
import net.hectus.neobb.util.component
import kotlin.reflect.KClass

abstract class Shop(val player: NeoPlayer) {
    companion object {
        fun turn(clazz: KClass<out Turn<*>>, player: NeoPlayer): Turn<*> {
            return if (clazz.simpleName?.contains("warp", true) == true) {
                clazz.java.getConstructor(PlacedStructure::class.java, NeoPlayer::class.java).newInstance(null, player)
            } else {
                clazz.constructors.first().call(null, null, player)
            }
        }
    }

    val loreBuilder: ItemLoreBuilder = player.game.info.loreBuilder.java.getConstructor().newInstance()
    val dummyTurns: List<Turn<*>> = player.game.info.turns.map { turn(it, player) }

    fun syncedOpen() {
        player.inventory.sync()
        open()
    }

    abstract fun open()

    fun done() {
        player.inventory.shopDone = true
        if (player.game.allShopDone()) {
            player.game.start()
        } else {
            player.sendMessage(player.locale().component("shop.wait", color = Colors.EXTRA))
        }
    }

    protected fun content(items: LinkedHashMap<ItemStack, Turn<*>>, consumer: (Map<ItemStack, Turn<*>>, ItemStack) -> Boolean): List<Item> {
        return items.sequencedKeySet().map { i -> Items.ClickItem(i) { _, e -> when (e.click) {
            ClickType.LEFT -> consumer(items, i)
            ClickType.RIGHT -> repeat(3) {
                if (!consumer(items, i))
                    return@ClickItem
            }
            ClickType.SHIFT_LEFT, ClickType.SHIFT_RIGHT -> repeat(player.game.info.deckSize) {
                if (!consumer(items, i))
                    return@ClickItem
            }
            else -> e.isCancelled = true
        } } }
    }
}
